runtime/interpreter/value.go

```diff
@@ -594,14 +595,20 @@ type StringValue struct {
 	graphemes *uniseg.Graphemes
 }
 
-func NewStringValue(str string) *StringValue {
+func NewUnmeteredStringValue(str string) *StringValue {
 	return &StringValue{
 		Str: str,
 		// a negative value indicates the length has not been initialized, see Length()
 		length: -1,
 	}
 }
 
+func NewStringValue(interpreter *Interpreter, memoryUsage MemoryUsage, stringConstructor func() string) *StringValue {
```

Comment:
If I understand correctly, the API requires the memory usage to be computed before the string is constructed so that we can elect to abort execution and not construct the string if we would go over our allotted memory. However, this API also complicates the creation of strings, and will ultimately complicate the construction of all kinds of values once we extend metering to everything else. 

Would it be possible to optimistically proceed with the creation of the string here (or int or other primitives) and check the used memory at the end, and then abort if that puts us over the limit? We might need to define a max string size or int size that is also checked to ensure we don't run out of memory as a result of the creation of a single value, but this would result in a much simpler implementation.

Reply:
Yes, I tried to design the API in such a way that the upfront memory usage calculation is "enforced" as much as possible: The constructor requires the upfront calculated measuring, and the allocation is performed lazily, after metering. I agree that this makes the construction of values slightly more work, but every allocation has the potential for an unmetered code path, so IMHO the extra effort required is worth the benefit.

I agree that it would be even better if we could devise an API that is simpler and still has the same effect.